### PR TITLE
Avoid waiting on validation during rolling update for inapplicable instance groups

### PR DESCRIPTION
**Pull request by bharath-123**
When an error occurs during cluster validation during a rolling update, the validation phase waits for a certain period of time before continuing. 

We can avoid this wait for certain errors inapplicable to an instance group whose rolling update occurring. eg: When a node in a different instance group is terminated leading to a error regarding instance group target size being lesser than expected. For such errors we do not want rolling update on the instance group to wait for errors on inapplicable instance groups.

Fixes [https://github.com/kubernetes/kops/issues/10009](https://github.com/kubernetes/kops/issues/10009)

Really appreciate the review comments and discussions. Working on this PR so far has been a great experience for me.

### CODE REVIEW

File: pkg/instancegroups/instancegroups.go [hunk 1]
```diff
@@ -477,14 +478,32 @@ func (c *RollingUpdateCluster) validateClusterWithTimeout(validateCount int) err
...
 		successCount = 0
 
-		// Wait before retrying
-		// TODO: Should we check if we have enough time left before the deadline?
-		time.Sleep(c.ValidateTickDuration)
+		// Wait before retrying in some cases
+		// TODO: Should we check if we have enough time left before the deadline? 
+		if shouldWaitBeforeRetry(result.Failures, group) {
+			time.Sleep(c.ValidateTickDuration)
+		}
```
Comment (johngmyers) [inline pkg/instancegroups/instancegroups.go:485]: There's a logic error here: when shouldWaitBeforeRetry fails, instead of letting the validation succeed and proceeding with updating the next node it instead goes into a fast spinloop.

When all failures are inapplicable it should go through the code block that increments `successCount`.

File: pkg/validation/validate_cluster_test.go [hunk 2]
```diff
@@ -154,22 +155,24 @@ func Test_ValidateCloudGroupMissing(t *testing.T) {
 			Kind:    "InstanceGroup",
 			Name:    "node-1",
 			Message: "InstanceGroup \"node-1\" is missing from the cloud provider",
+			InstanceGroup: instanceGroup,
```
Comment (johngmyers) [inline pkg/validation/validate_cluster_test.go:158]: Nit: you could have made a smaller change by using `instanceGroups[0]`.

**Comment by bharath-123** [comment]
Seems like some static checks and format tests are failing. Should have run gofmt and a bunch of other static checks before raising the PR. Will do so now.

**Review by johngmyers — CHANGES_REQUESTED** [review]
You'll need to run `make gofmt` and fix that duplicate import in order to fix those failed tests.

A useful unit test for rolling update would be to mock in a validation failure for a nodes group that is not selected for update and verify that the rolling update completes regardless.

The validation unit tests should be updated to assert that the instanceGroup field is filled in appropriately.

In `collectPodFailures`, failures for pods of priority `system-node-critical` should be marked as being specific to the pod's node's group. These are typically things like CNI DaemonSets. Pods that are `system-cluster-critical`, on the other hand, are cluster-level failures.

**Comment by johngmyers** [comment]
If the `system-node-critical` feature is too complicated, it can be done as a follow-up PR.

**Comment by bharath-123** [comment]
@johngmyers I was not able to figure out how to infer the InstanceGroup for a pod in collectPodFailures? The pod has a NodeName field in it's podspec. But I was not really sure how we could infer the InstanceGroup for a pod from that?

**Comment by bharath-123** [comment]
I will follow this up with the unit test and the code changes that you have suggested. Thank you.

**Comment by johngmyers** [comment]
It would be better to edit the commit stream so that the formatting and staticcheck issues were fixed in their original commits.

I like the separation between the first and second commits. The third commit should be combined with the second as it is part of the same logical change.

**Comment by bharath-123** [comment]
> If the `system-node-critical` feature is too complicated, it can be done as a follow-up PR.

Shouldn't be difficult if we can figure out how to infer the instanceGroup from a pod given its structure.

**Comment by bharath-123** [comment]
> It would be better to edit the commit stream so that the formatting and staticcheck issues were fixed in their original commits.
> 
> I like the separation between the first and second commits. The third commit should be combined with the second as it is part of the same logical change.

yes I agree with this. I will anyways be doing a whole bunch of changes now. So will revise the commit stream.

File: pkg/instancegroups/instancegroups.go (hunk 1 above)
Comment (bharath-123) [reply pkg/instancegroups/instancegroups.go:485]: One question, even if the failures are inapplicable and we skip the wait period. We definitely want to log the failures right? I think we would need to change the log messages here if I am right. If the validation fails and the failures are inapplicable, showing the failure log messages and a cluster validated messages would be very unintuitive for the sys admins. 
I am planning to copy this block of code and add in the else stmt if shouldWaitBeforeRetry fails. I will change up the log messages to indicate that there **were** failures but they were ignored because they are in a different instance group.
```
			successCount++
			if successCount >= validateCount {
				klog.Info("Cluster validated.")
				return nil
			}
			klog.Infof("Cluster validated; revalidating in %s to make sure it does not flap.", c.ValidateSuccessDuration)
			time.Sleep(c.ValidateSuccessDuration)
			continue
```

**Comment by johngmyers** [comment]
`validateNodes` could construct a map of nodes to groups. `collectPodFailures` already constructs a map of IP address to node, so it could then use that map of nodes to groups to go from pod to group.

**Comment by johngmyers** [comment]
The rolling update code should probably remove the inapplicable failures from what it logs.

File: pkg/validation/validate_cluster.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ import (
 
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/tools/pager"
+	kopsapi "k8s.io/kops/pkg/apis/kops"
```
Comment (justinsb) [inline pkg/validation/validate_cluster.go:28]: This should already be imported as kops below, I believe

**Comment by johngmyers** [comment]
Okay, not sure why kube-apiserver defaults that off. An admission controller would be more in line with current practice and would allow policies more nuanced than "prohibit everything".

**Comment by bharath-123** [comment]
@johngmyers I was working on this? Hard to manage this and office work, so I mostly work on the weekends :/ Any reason why this was closed?

**Comment by bharath-123** [comment]
I did not understand your last comment?

**Comment by johngmyers** [comment]
Sorry, wrong PR.

File: pkg/validation/validate_cluster.go (hunk 3 above)
Comment (bharath-123) [reply pkg/validation/validate_cluster.go:28]: yes @justinsb will be doing this.

**Comment by bharath-123** [comment]
> `validateNodes` could construct a map of nodes to groups. `collectPodFailures` already constructs a map of IP address to node, so it could then use that map of nodes to groups to go from pod to group.

Great idea. I ll do this in a follow up PR tho. All the pod failures caught during validate cluster phase are system critical anyways. By not adding the instanceGroup field to them, I am not ignoring the errors.

**Comment by bharath-123** [comment]
> The rolling update code should probably remove the inapplicable failures from what it logs.

Makes sense. I am working on a different implementation of this feature now.

**Comment by bharath-123** [comment]
@johngmyers @justinsb @hakman @olemarkus @zetaab do have a look at this. This is a complete implementation with unit tests added for rolling update. Do review this.

**Comment by bharath-123** [comment]
There are a couple of unused imports and unrelated comments which I had added when I was trying a different implementation. I ll remove them now.

**Comment by bharath-123** [comment]
gentle ping @johngmyers @justinsb @hakman

**Comment by rifelpet** [comment]
@bharath-123 Feel free to remove `[WIP]` from the PR title if this is ready for review.

**Comment by bharath-123** [comment]
> @bharath-123 Feel free to remove `[WIP]` from the PR title if this is ready for review.

done

File: pkg/instancegroups/rollingupdate_test.go [hunk 4]
```diff
@@ -105,6 +105,30 @@ func (*erroringClusterValidator) Validate() (*validation.ValidationCluster, erro
...
+// simulates failures in a specific node group in the map of instance groups
+type instanceGroupNodeSpecificErrorClusterValidator struct {
+	Groups    map[string]*cloudinstances.CloudInstanceGroup
+	NodeGroup string
+}
+
+func (igErrorValidator *instanceGroupNodeSpecificErrorClusterValidator) Validate() (*validation.ValidationCluster, error) {
+	instanceGroup := igErrorValidator.Groups[igErrorValidator.NodeGroup].InstanceGroup
+	if instanceGroup.IsMaster() || instanceGroup.IsBastion() {
+		instanceGroup = nil
```
Comment (johngmyers) [inline pkg/instancegroups/rollingupdate_test.go:117]: This is testing the mock, not the code under test.

**Review by johngmyers — CHANGES_REQUESTED** [review]
I don't see where in the code failures specific to a master IG cause rolling update to be blocked from proceeding.

From comments it seems you intended to put this logic in the validation code. I think it might be simpler to put it in the rolling update code that examines the returned list of failures.

File: pkg/instancegroups/instancegroups.go [hunk 5]
```diff
@@ -477,14 +478,26 @@ func (c *RollingUpdateCluster) validateClusterWithTimeout(validateCount int) err
...
-		// Wait before retrying
+		// Wait before retrying in some cases
 		// TODO: Should we check if we have enough time left before the deadline?
 		time.Sleep(c.ValidateTickDuration)
 	}
 
 	return fmt.Errorf("cluster did not validate within a duration of %q", c.ValidationTimeout)
 }
 
+func isNotRelatedInstanceGroupError(failures []*validation.ValidationError, group *cloudinstances.CloudInstanceGroup) bool {
```
[the remark this answers is not in the document]
Comment (bharath-123) [reply pkg/instancegroups/instancegroups.go:489]: yes. makes sense. Will do that.

File: pkg/instancegroups/instancegroups.go [hunk 6]
```diff
@@ -445,7 +446,7 @@ func (c *RollingUpdateCluster) validateClusterWithTimeout(validateCount int) err
 	for {
 		// Note that we validate at least once before checking the timeout, in case the cluster is healthy with a short timeout
 		result, err := c.ClusterValidator.Validate()
-		if err == nil && len(result.Failures) == 0 {
+		if err == nil && (len(result.Failures) == 0 || isNotRelatedInstanceGroupError(result.Failures, group)) {
```
[the remark this answers is not in the document]
Comment (bharath-123) [reply pkg/instancegroups/instancegroups.go:449]: yes. Will remove it.

File: pkg/validation/validate_cluster.go [hunk 7]
```diff
@@ -49,6 +49,10 @@ type ValidationError struct {
 	Kind    string `json:"type,omitempty"`
 	Name    string `json:"name,omitempty"`
 	Message string `json:"message,omitempty"`
+	// optional field to indicate which instance group this validation error is coming from
+	// if InstanceGroup is nil, then the validation error is either a cluster wide error
+	// or an error in the master instancegroup which indicates a cluster wide error
```
[the remark this answers is not in the document]
Comment (bharath-123) [reply pkg/validation/validate_cluster.go:54]: Acked these comment suggestions. They are indeed more helpful.

File: pkg/instancegroups/rollingupdate_test.go [hunk 8]
```diff
@@ -105,6 +105,30 @@ func (*erroringClusterValidator) Validate() (*validation.ValidationCluster, erro
...
 }
 
+// simulates failures in a specific node group in the map of instance groups
+type instanceGroupNodeSpecificErrorClusterValidator struct {
+	Groups    map[string]*cloudinstances.CloudInstanceGroup
+	NodeGroup string
+}
+
+func (igErrorValidator *instanceGroupNodeSpecificErrorClusterValidator) Validate() (*validation.ValidationCluster, error) {
+	instanceGroup := igErrorValidator.Groups[igErrorValidator.NodeGroup].InstanceGroup
```
[the remark this answers is not in the document]
Comment (bharath-123) [reply pkg/instancegroups/rollingupdate_test.go:115]: alright. Will do that. It makes sense.

File: pkg/validation/validate_cluster.go [hunk 9]
```diff
@@ -238,6 +243,9 @@ func (v *ValidationCluster) collectPodFailures(ctx context.Context, client kuber
 		if pod.Status.Phase == v1.PodSucceeded {
 			return nil
 		}
+
+		// pod validationErrors usually do not have instanceGroup field as all pod errors caught
+		// are system critical pods.
```
[the remark this answers is not in the document]
Comment (bharath-123) [reply pkg/validation/validate_cluster.go:248]: I inferred this from line 240 containing the following code:
```
		if priority != "system-cluster-critical" && priority != "system-node-critical" {
			return nil
		}
```
But I think I wrongly read system-node-critical. Need to brainstorm on how to get the InstanceGroup for the pods which doesn't seem to be straightforward.

**Comment by bharath-123** [comment]
> I don't see where in the code failures specific to a master IG cause rolling update to be blocked from proceeding.
> 
> From comments it seems you intended to put this logic in the validation code. I think it might be simpler to put it in the rolling update code that examines the returned list of failures.



> I don't see where in the code failures specific to a master IG cause rolling update to be blocked from proceeding.
> 
> From comments it seems you intended to put this logic in the validation code. I think it might be simpler to put it in the rolling update code that examines the returned list of failures.

In the ValidationError struct, If the InstanceGroup field is nil then this failure is a cluster wide failure for which rolling update should indeed be blocked. For all validation failures in master IG, the InstanceGroup field is nil. isNotRelatedInstanceGroupError ( Which I will change the name of ) returns false if InstanceGroup field is nil which causes rolling update to be blocked. 

It would definitely be cleaner if we can have InstanceGroup field filled up for all failures. I mainly did this because it would be a bit difficult to identify the groups for the pods. I ll think more about this and get back.

File: pkg/instancegroups/rollingupdate_test.go (hunk 4 above)
Comment (bharath-123) [reply pkg/instancegroups/rollingupdate_test.go:117]: Would it be better if we include the Validate logic in the Validate mock? I thought just mocking the failures should have been enough to test this. Since we are mostly concerned about the failures return by Validate() and not how it returns it. Am I understanding something wrongly?

File: pkg/validation/validate_cluster.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (johngmyers) [reply pkg/validation/validate_cluster.go:248]: As mentioned previously, this can be deferred to a later PR.

File: pkg/instancegroups/rollingupdate_test.go (hunk 4 above)
Comment (johngmyers) [reply pkg/instancegroups/rollingupdate_test.go:117]: If it was the rolling update code that was responsible for ensuring failures for IGs of role master still blocked rolling update, then testing that in the rolling updates code would make sense. In the current design, where it is the validation code that is responsible for that logic, that logic should be tested by the validation unit tests, not the rolling update unit tests.

**Comment by johngmyers** [comment]
> For all validation failures in master IG, the InstanceGroup field is nil.

I did not see code for ensuring the `InstanceGroup` field is `nil` for failures in IGs of role master.

File: pkg/validation/validate_cluster.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (bharath-123) [reply pkg/validation/validate_cluster.go:248]: Alright. So we can leave the InstanceGroup field as nil for validation error for kind Pods.

File: pkg/instancegroups/rollingupdate_test.go (hunk 4 above)
Comment (bharath-123) [reply pkg/instancegroups/rollingupdate_test.go:117]: Fair enough. This makes sense. I ll push all logic to determine whether we should wait or not for validation to the rolling update code rather than the validation code. It would make things much easier and simpler

**Comment by bharath-123** [comment]
> > For all validation failures in master IG, the InstanceGroup field is nil.
> 
> I did not see code for ensuring the `InstanceGroup` field is `nil` for failures in IGs of role master.

Will be pushing all the logic to rolling update code now.

**Comment by bharath-123** [comment]
@johngmyers working on refining this. Just to confirm, for Pods and ComponentStatuses it's not straightforward to get the InstanceGroup. This will anyways be deferred to a future PR. 

For these cases, in the rolling update code we will be considering cases where `InstanceGroup == nil` to be a relevant InstanceGroup failure. Would I need to assert these cases in the validate_cluster tests? In the rolling update test, I ll include a test to fail rolling update if a validation error is returned if InstanceGroup is nil.

**Comment by bharath-123** [comment]
gentle ping on this @johngmyers . Thank you for your help and patience on this PR.
